perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java

```diff
@@ -2718,6 +2720,155 @@ public void updateSponsorshipValidity(PerunSession sess, Member sponsoredMember,
+			for (Attribute attribute : memberAttributes) {
+				if (!getPerunBl().getAttributesManagerBl().isVirtAttribute(sess, attribute) &&
+					!getPerunBl().getAttributesManagerBl().isCoreAttribute(sess, attribute))
+					getPerunBl().getAttributesManagerBl().setAttribute(sess, targetMember, attribute);
+			}
+
+			// set member-group attributes
+			// add 'members' group to directGroups to set its member-group attributes too
+			directGroups.add(getPerunBl().getGroupsManagerBl().getGroupByName(sess, vo, VosManager.MEMBERS_GROUP));
+			for (Group group : directGroups) {
+				List<Attribute> groupMemberAttributes = getPerunBl().getAttributesManagerBl().getAttributes(sess, sourceMember, group);
+				for (Attribute attribute : groupMemberAttributes) {
+					if (!getPerunBl().getAttributesManagerBl().isVirtAttribute(sess, attribute)) {
+						getPerunBl().getAttributesManagerBl().setAttribute(sess, targetMember, group, attribute);
+					}
+				}
+			}
+
+			// set member-resource attributes
+			List<Resource> resources = getPerunBl().getResourcesManagerBl().getResources(sess, vo);
+			for (Resource resource : resources) {
+				List<Attribute> memberResourceAttributes = getPerunBl().getAttributesManagerBl().getAttributes(sess, sourceMember, resource);
+				for (Attribute attribute : memberResourceAttributes) {
+					if (!getPerunBl().getAttributesManagerBl().isVirtAttribute(sess, attribute)) {
+						getPerunBl().getAttributesManagerBl().setAttribute(sess, targetMember, resource, attribute);
+					}
+				}
+			}
+		} catch (WrongAttributeValueException | WrongAttributeAssignmentException | WrongReferenceAttributeValueException |
+				 MemberGroupMismatchException | MemberResourceMismatchException | GroupNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		// set status from old member
+		// first validate member to prevent MemberNotValidYetException later
+		try {
+			this.validateMember(sess, targetMember);
+		} catch (WrongAttributeValueException | WrongReferenceAttributeValueException e) {
+			throw new InternalErrorException(e);
+		}
+		if (sourceMember.getStatus() != Status.VALID) {
+			try {
+				this.setStatus(sess, targetMember, sourceMember.getStatus());
+			} catch (WrongAttributeValueException | WrongReferenceAttributeValueException | MemberNotValidYetException e) {
+				throw new InternalErrorException(e);
+			}
+		}
+
+		// move all applications
+		getMembersManagerImpl().moveMembersApplications(sess, sourceMember, targetMember);
+
+		// move sponsorships in the VO
+		// move sponsorships where sourceUser is sponsor
+		List<Member> sponsoredMembers = this.getSponsoredMembers(sess, vo, sourceUser);
+		for (Member sponsoredMember : sponsoredMembers) {
+			try {
+				Sponsorship sponsorship = this.getSponsorship(sess, sponsoredMember, sourceUser);
+				// first add sponsorship to targetUser
+				this.sponsorMember(sess, sponsoredMember, targetUser, sponsorship.getValidityTo());
```

Comment:
If the targetUser doesn't have the SPONSOR role in the given Vo, we want to assign it to him so this would not fail.

Reply:
Done.